internal/campaigns/executor.go

```diff
@@ -302,8 +319,26 @@ func (x *executor) do(ctx context.Context, task *Task) (err error) {
 	return
 }
 
-func (x *executor) updateTaskStatus(task *Task, status *TaskStatus) {
-	x.tasks.Store(task, status)
+func (x *executor) updateTaskStatus(task *Task, update func(status *TaskStatus)) {
+	x.statusesMu.Lock()
+	defer x.statusesMu.Unlock()
+
+	status, ok := x.statuses[task]
+	if ok {
+		update(status)
+	}
+}
+
+func (x *executor) LockedTaskStatuses(callback func([]*TaskStatus)) {
+	x.statusesMu.RLock()
+	defer x.statusesMu.RUnlock()
+
+	var s []*TaskStatus
+	for _, status := range x.statuses {
+		s = append(s, status)
+	}
```

Comment:
why the copy of a pointer slice?

Reply:
Because `x.statuses` is a map and not a slice and I don't want to expose the `*Task` (key of the map) to the consumers.

Reply:
Ah, I didn't see that. Makes sense